explorer/templates.go

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2017, The dcrdata developers
+// See LICENSE for details.
+
+package explorer
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"math"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg"
+	humanize "github.com/dustin/go-humanize"
+)
+
+type pageTemplate struct {
+	file     string
+	template *template.Template
+}
+
+type templates struct {
+	templates map[string]pageTemplate
+	defaults  []string
+	folder    string
+	helpers   template.FuncMap
+}
+
+func newTemplates(folder string, defaults []string, helpers template.FuncMap) templates {
+	var defs []string
+	for _, file := range defaults {
+		defs = append(defs, filepath.Join(folder, file+".tmpl"))
+	}
+	return templates{
+		templates: make(map[string]pageTemplate),
+		defaults:  defs,
+		folder:    folder,
+		helpers:   helpers,
+	}
+}
+
+func (t templates) addTemplate(name string) error {
+	fileName := filepath.Join(t.folder, name+".tmpl")
+	files := append(t.defaults, fileName)
+	temp, err := template.New(name).Funcs(t.helpers).ParseFiles(files...)
+	if err == nil {
+		t.templates[name] = pageTemplate{
```

Comment:
value receiver OK in this function?

Reply:
I don't follow

Reply:
I'm referring to the use of `(t templates)` instead of `(t *templates)`

Reply:
Oh, yeah... a pointer receiver would be better since it could mutate `t.templates`. Changed all `t` receivers for consistency.

Reply:
BTW, I think this would have worked fine because of the "true" nature of maps in go, but I think it was right to change it to a pointer receiver.

FYI https://dave.cheney.net/2017/04/30/if-a-map-isnt-a-reference-variable-what-is-it